dbms/src/Interpreters/Expand2.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Interpreters/Expand2.h>
+
+namespace DB
+{
+
+Expand2::Expand2(ExpressionActionsPtrVec projections_actions, NamesWithAliasesVec projections)
+    : leveled_projections_actions(projections_actions)
+    , leveled_alias_projections(projections)
+{
+}
+
+Block Expand2::next(const Block & block_cache, size_t i_th_project)
+{
+    /// step1: clone a new block
+    ColumnsWithTypeAndName cloned_columns_and_type;
+    auto num_columns = block_cache.columns();
+    MutableColumns res(num_columns);
+    for (size_t i = 0; i < num_columns; ++i)
+    {
+        res[i] = block_cache.getColumns()[i]->cloneResized(block_cache.rows());
```

Comment:
For the columns that must be null, I think we should not clone their full content,

Reply:
make sense.
we should dig out which col in this level is projected as null, then, do something elimination of copy work here. For now, multi_level_actions is seen as black box here, no too much analyzing work, just use `level_action.execute(block)` to get more simplicity

Reply:
> make sense. we should dig out which col in this level is projected as null (varies from levels), then, do something elimination of copy work here. For now, multi_level_actions is seen as black box here, no too much analyzing work, just use `level_action.execute(block)` to get more simplicity

projection action will add null attribute on some columns in L46. I think we can record which columns are null before and after executing. By comparing the difference we can know which columns are the columns that is projected to null and we can avoid the cloning on these columns.

Reply:
Please add todo for further performance optimization.